### ISSUE TITLE
Failed to download

### ISSUE THREAD

**Issue by sejooning** [EGA-archive/ega-download-client#97]
Dear ega-download client developers,

I  got some errors to download EGA dataset using pyEGA3 client. 
Error message is like below.

[2020-11-12 12:31:14 +0900] Download starting [using 1 connection(s)]...
^M  0%|          | 0.00/31.8G [00:00<?, ?B/s][2020-11-12 12:31:14 +0900] Starting new HTTPS connection (1): ega.ebi.ac.uk
^M  0%|          | 0.00/31.8G [01:09<?, ?B/s]
[2020-11-12 12:32:23 +0900] __exit__
Traceback (most recent call last):
  File "../ega-download-client/pyega3/pyega3.py", line 494, in download_file_retry
    download_file(token, file_id, file_size, check_sum, num_connections, key, output_file)
  File "../ega-download-client/pyega3/pyega3.py", line 422, in download_file
    for part_file_name in executor.map(download_file_slice_, params):
  File "/usr/lib/python3.5/concurrent/futures/_base.py", line 556, in result_iterator
    yield future.result()
  File "/usr/lib/python3.5/concurrent/futures/_base.py", line 405, in result
    return self.__get_result()
  File "/usr/lib/python3.5/concurrent/futures/_base.py", line 357, in __get_result
    raise self._exception
  File "/usr/lib/python3.5/concurrent/futures/thread.py", line 55, in run
    result = self.fn(*self.args, **self.kwargs)
  File "../ega-download-client/pyega3/pyega3.py", line 296, in download_file_slice_
    return download_file_slice(*args)
  File "../ega-download-client/pyega3/pyega3.py", line 280, in download_file_slice
    with requests.get(url, headers=headers, stream=True) as r:
AttributeError: __exit__

Any helps would be appreciated.

**Comment by malloryfreeberg** [comment]
Hi @sejooning,

Thank you for bringing this issue to our attention. Can you please confirm what version of pyega3 you are using? If you're not using the latest version, can you please update to the latest version of pyega3 - 3.4.0 - and then retry your download?

**Comment by sejooning** [comment]
Hi malloryfreeberg, 

Thank you for your reply. I used pyega 3.4.0 version.
I am running pyega3 using unbuntu linux, but I chose the debian version for installing pyega3 as the provided install script is only for the red_hat and debian versions of Linux. I think this can cause problems. Do you have an example of installing pyega3 on Unbuntu Linux?

**Comment by malloryfreeberg** [comment]
Hi @sejooning 

It appears that your installation of pyega3 is OK, since you made it past the authorisation step and to the "Download starting [using 1 connection(s)]..." step. To confirm that pyega3 is installed correctly, can you please run the following:

1. Fresh install of latest pyega3:
```
sudo pip3 install pyega3
```

2. Test listing available datasets with the "test" (-t) account:
```
pyega3 -t datasets
```

3. Test listing files available in a test dataset:
```
pyega3 -t files EGAD00001003338
```

4. Test downloading a file from the test dataset:
```
pyega3 -t fetch EGAF00001775036
```

These are the same steps as listed for testing pyega3 installation in the [README](https://github.com/EGA-archive/ega-download-client#testing-pyega3-installation).

We have confirmed these steps to work for Ubuntu Linux. Please note there might be a delay in starting the download, so please be patient and do not exit early. Do let me know how you get on after completing the indicated steps.

**Comment by malloryfreeberg** [comment]
Hi @sejooning 

Just checking in to see whether you have been able to download the files you were after?

**Comment by fklirono** [comment]
Hello, I have been having a similar issue when running pyega3 (3.4.0) behind our network's web-proxy. It seems this is a problem that only my machine experiences, which makes it extremely annoying and difficult to identify. Other users do not seem to have this problem! Here's the output when attempting to download the test data:

.........................................................................................................................................................................................................................
[2020-11-27 22:50:46 +0100] pyEGA3 - EGA python client version 3.4.0 (https://github.com/EGA-archive/ega-download-client)
[2020-11-27 22:50:46 +0100] Parts of this software are derived from pyEGA (https://github.com/blachlylab/pyega) by James Blachly
[2020-11-27 22:50:46 +0100] Python version : 3.5.2
[2020-11-27 22:50:46 +0100] OS version : Linux #126~16.04.1-Ubuntu SMP Wed Oct 21 13:48:05 UTC 2020
[2020-11-27 22:50:46 +0100] Server URL: https://ega.ebi.ac.uk:8052/elixir/data
[2020-11-27 22:50:46 +0100] Session-Id: 458524637
[2020-11-27 22:50:46 +0100] Starting new HTTPS connection (1): ega.ebi.ac.uk
[2020-11-27 22:50:47 +0100] "POST /ega-openid-connect-server/token HTTP/1.1" 200 None
[2020-11-27 22:50:47 +0100] 
[2020-11-27 22:50:47 +0100] Authentication success for user 'ega-test-data@ebi.ac.uk'
[2020-11-27 22:50:47 +0100] Starting new HTTPS connection (1): ega.ebi.ac.uk
[2020-11-27 22:50:54 +0100] "GET /elixir/data/metadata/files/EGAF00001775036 HTTP/1.1" 200 None
[2020-11-27 22:50:54 +0100] Request URL : https://ega.ebi.ac.uk:8052/elixir/data/metadata/files/EGAF00001775036
[2020-11-27 22:50:54 +0100] Response    :
 {
    "fileName": "EGAZ00001314787/ENCFF284YOU.bam.bai.cip",
    "fileSize": 4804928,
    "fileStatus": "available",
    "unencryptedChecksumType": "MD5",
    "unencryptedChecksum": "3b89b96387db5199fef6ba613f70e27c",
    "displayFilePath": "EGAZ00001314787",
    "fileId": "EGAF00001775036",
    "datasetId": "EGAD00001003338",
    "displayFileName": "ENCFF284YOU.bam.bai"
}
[2020-11-27 22:50:54 +0100] Starting new HTTPS connection (1): ega.ebi.ac.uk
[2020-11-27 22:50:55 +0100] "POST /ega-openid-connect-server/token HTTP/1.1" 200 None
[2020-11-27 22:50:55 +0100] 
[2020-11-27 22:50:55 +0100] Authentication success for user 'ega-test-data@ebi.ac.uk'
[2020-11-27 22:50:55 +0100] File Id: 'EGAF00001775036'(4804928 bytes).
[2020-11-27 22:50:55 +0100] Output file:'EGAF00001775036/ENCFF284YOU.bam.bai'
[2020-11-27 22:50:55 +0100] Total space : 1833.66 GiB
[2020-11-27 22:50:55 +0100] Used space : 280.94 GiB
[2020-11-27 22:50:55 +0100] Free space : 1459.55 GiB
[2020-11-27 22:50:55 +0100] Download starting [using 1 connection(s)]...
[2020-11-27 22:50:55 +0100] Starting new HTTPS connection (1): ega.ebi.ac.uk
[2020-11-27 22:51:03 +0100] "GET /elixir/data/files/EGAF00001775036?destinationFormat=plain HTTP/1.1" 206 4804912
[2020-11-27 22:51:03 +0100] __exit__
Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/pyega3/pyega3.py", line 494, in download_file_retry
    download_file(token, file_id, file_size, check_sum, num_connections, key, output_file)
  File "/usr/local/lib/python3.5/dist-packages/pyega3/pyega3.py", line 422, in download_file
    for part_file_name in executor.map(download_file_slice_, params):
  File "/usr/lib/python3.5/concurrent/futures/_base.py", line 556, in result_iterator
    yield future.result()
  File "/usr/lib/python3.5/concurrent/futures/_base.py", line 405, in result
    return self.__get_result()
  File "/usr/lib/python3.5/concurrent/futures/_base.py", line 357, in __get_result
    raise self._exception
  File "/usr/lib/python3.5/concurrent/futures/thread.py", line 55, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/usr/local/lib/python3.5/dist-packages/pyega3/pyega3.py", line 296, in download_file_slice_
    return download_file_slice(*args)
  File "/usr/local/lib/python3.5/dist-packages/pyega3/pyega3.py", line 280, in download_file_slice
    with requests.get(url, headers=headers, stream=True) as r:
AttributeError: __exit__
.........................................................................................................................................................................................................................

As you can see, the connection to the server is established. The metadata are downloaded, however, when it comes to downloading the actual data all child processes exit! 

Any help would be greatly appreciated to solve this issue!

**Comment by malloryfreeberg** [comment]
Hi @fklirono. Thank you for reporting this issue. As this is an issue related to your specific user account, if you are still experiencing this behavior please email EGA Helpdesk at helpdesk@ega-archive.org with this issue and attach the output log produced by pyega3 (pyega3_output.log).

@sejooning - If you are also still experiencing this behavior, please contact EGA Helpdesk team as described above.

For now, I will close this issue as this is a service issue that the EGA Helpdesk team is best suited to provide further assistance.

**Comment by malloryfreeberg** [comment]
Hi @fklirono . You are correct. I had confused this ticket with another ticket reporting 500 server errors. We have added this suggestion to our Helpdesk SOPs for assisting users using pyega3.

Also, thank you for the suggestion for improving the pyega3 download client. Indeed, there should be a more user-friendly way to flag and/or prevent this particular issue. We have scheduled work to address this issue in our internal ticketing system, and we appreciate the feedback.
